rfcs/20230517-forward-compatibility-testing.md

```diff
@@ -0,0 +1,99 @@
+check backward compatibility by versioning and serializing
+`stablehlo_legalize_to_vhlo.mlir`, and need something similar for forward
+compatibility. This document makes the following proposal:
+
+**Proposal:** Add tests to compare `stablehlo_legalize_to_vhlo.0_X_0.mlir`
+serialized at HEAD with existing [`stablehlo_legalize_to_vhlo.0_X_0.mlir.bc`](https://github.com/openxla/stablehlo/blob/main/stablehlo/tests/stablehlo_legalize_to_vhlo.0_9_0.mlir.bc)
+test files.
+
+## Requirements
+
+Forward compatibility testing has the following requirements, to ensure that it
+is both useful and practical:
+
+* **R1.** Tests must detect changes in StableHLO that break forward compatibility.
+* **R2.** Tests must detect LLVM revision bumps that break forward compatibility.
+* **R3.** Tests must scale with releases and execute in a reasonable amount
+of time.
+
+## Proposed Design: Statically Test Forward Compatibility
+
+In this design, forward incompatibilities are detected by serializing
+`stablehlo_legalize_to_vhlo.mlir` at head and comparing it to a known-good file
+serialized serialized at previous releases. This approach operates on the
+assumption that at HEAD we are able to produce a byte-identical serialized
+artifact, excluding custom header info and debug locations.
+
+Detecting forward incompatiblities can be accomplished by adding the following
+`RUN` line to the versioned [`stablehlo_legalize_to_vhlo.0_X_0.mlir`](https://github.com/search?q=repo%3Aopenxla%2Fstablehlo+path%3A**%2Fstablehlo_legalize_to_vhlo.0_*&type=code)
+test files.
+
+```bash
+diff %s.bc <(stablehlo-translate --serialize --target=0.X.0)
+```
+
+This test should be run in CI on all PRs to prevent merging incompatible changes
+in StableHLO (_R1_) and LLVM revision bumps (_R2_).
+
+This will not work as is because `bytecode::producer`, `op::location` and
+`block_argument::location` fields may change between when the original bytecode
+file has been produced and when the test runs at HEAD. To account for that,
+we can strip this information before comparing the payloads.
+
+The StableHLO process for [Contributing Incompatible Changes](https://github.com/openxla/stablehlo/blob/main/docs/vhlo.md#add-versioned-serialization-test)
+requires `stablehlo_legalize_to_vhlo.mlir` to be serialized for each
+compatibility breaking change, which means that we can test forward
+compatibility between HEAD and each minor release of StableHLO. Given that
+all changes to the StableHLO opset, as well as changes to serialization
+machinery like targeting a newer version of the MLIR bytecode format, require
+bumping the minor version, there is a low chance of causing forward
+incompatibilities between different patch versions of StableHLO (i.e.
+`0.10.1 → 0.10.2`). Ensuring compatibility between HEAD and all minor releases
+(`0.9.0`, `0.10.0`, etc.) should provide thorough enough coverage of the two
+mentioned sources of forward incompatibilities.
```

Comment:
How do we define "thorough enough"? The current version of [compatibility.md](https://github.com/openxla/stablehlo/blob/main/docs/compatibility.md) is pretty strict about these things.

Reply:
Changed to "full coverage" let's resolve the discussion of full coverage in https://github.com/openxla/stablehlo/pull/1498#discussion_r1206067372, resolving this comment.